src/interceptors/ClientRequest/MockHttpSocket.ts

```diff
@@ -205,6 +207,16 @@ export class MockHttpSocket extends MockSocket {
     this.mockConnect()
     this.responseType = 'mock'
 
+    // Exhaust the "socket.write()" callbacks to
+    // transition the Writable into the right state
+    // (i.e. "request body written" state). There's nowhere
+    // to actually write the data so we disregard it.
+    // It has been used by the request parser to construct
+    // a Fetch API Request instance representing this request.
+    for (const [_, __, writeCallback] of this.writeBuffer) {
+      writeCallback?.()
```

Comment:
The actual fix. 

When ClientRequest writes the data to the socket, the `.write()` method of the MockSocket just buffers the chunks in `this.writeBuffer`. It never actually does any writes (we cannot do that until we know if the connection is possible) and so it never called any write callbacks (such as the `onFinish` callback).

Reply:
Maybe we now can/should return false in the `socker.write`?

Reply:
Now I see we are calling the write callback after the response, which causes a deadlock:

```js
const http = require('http')
const sinon = require('sinon')
const { ClientRequestInterceptor } = require('@mswjs/interceptors/ClientRequest')

const interceptor = new ClientRequestInterceptor()

interceptor.on('request', async function rootListener({ request }) {
  console.log('before');
  await request.arrayBuffer()
  console.log('never get here');
  request.respondWith(new Response('OK!'))
})
interceptor.apply()

const reqWriteCallback = sinon.spy()

const req = http.request(
  {
    host: 'example.com',
    method: 'POST',
    path: '/',
    port: 80,
  },
  res => {
    console.log(2)
    res.on('end', () => {
      console.log(3);
    })
    // Streams start in 'paused' mode and must be started.
    // See https://nodejs.org/api/stream.html#stream_class_stream_readable
    res.resume()
  },
)

req.write('mamma mia', null, () => {
  console.log(1);
  reqWriteCallback()
  req.end()
})
```

Reply:
Are you talking about the response _event_ when you mention "response"?

Reply:
I mean we call the `write` callback inside the `responseWith` function. sorry for the confusion.

Reply:
Got it! Thanks for clarifying. 

In your example, this is the order I'd expect:

```
1
before
never get here
2
3
```

Is that the expected order?

Reply:
> Is that the expected order?

TL;DR:  I think yes.

It's a design choice. If I understand you correctly, you would rather the whole request body to be ready for the user in the interceptor so that they can do `await request.arrayBuffer` without a fear of deadlock like this.
We can also say that the body isn't guaranteed to be ready, and they user needs to take care of this scenario by itself (I think they can with `request.arrayBuffer().then(..)`)

Reply:
Actually, not quite. The request event is emitted as soon as the request's headers are sent. So the request body may still be streaming. If the user decides to read it, whichever body reading method will return a Promise, and the entire request listener would have to wait for that promise. 

I wonder if that conflicts with the write callbacks. As in, in order to read the request body, the socket has to call the callbacks of .write(), and we are calling those only as a part of `respondWith()`. 

That shouldn't be the case though. The "request body" the user is reading in the request listener is our internal request body buffer where we buffer the chunks pushed.

Reply:
> That shouldn't be the case though. The "request body" the user is reading in the request listener is our internal request body buffer where we buffer the chunks pushed.

But the internal request body never ends because we never call the write callback, which ends the request body. (we never do: `this.requestStream.push(null)`)

Reply:
Not sure we should be doing this. Need to look more into this behavior but it's the actual request that calls `.end()` on the writable stream. Perhaps the issue is that we are not translating that to `this.requestStream.push(null)`.

Reply:
I've pushed a fix in 7be0242029db25f9d3638ffc4a4ea9f4777c73f9. 

Basically, if something attempts to read the request buffer (e.g. `await request.text()` in the interceptor), the `MockHttpSocket` will immediately start flushing all the write callbacks. 

If nothing reads the buffer, the write callbacks will be flushed before the mocked response starts streaming or before the original request is made.